secret/secret.go

```diff
@@ -64,48 +68,147 @@ func UpdateSecrets(s secretInterface, secrets *v1.Secret, create bool) {
 		if err != nil {
 			logFatal(err)
 		}
-		log.Println("Created `secret`")
+		log.Printf("Created `%s`\n", secrets.Name)
 	} else if util.IsDirty(secrets) {
 		util.MarkAsClean(secrets)
 		_, err := s.Update(secrets)
 		if err != nil {
 			logFatal(err)
 		}
-		log.Println("Updated `secret`")
+		log.Printf("Updated `%s`\n", secrets.Name)
+	}
+}
+
+func FindSecret(s secretInterface) (*v1.Secret, error) {
+	slist, err := s.List (metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	var best v1.Secret
+	for _, s := range slist.Items {
+		if (s.Name != SECRET_NAME) && !secretPattern.MatchString(s.Name) {
+			// We allow only
+			// - SECRET_NAME exactly
+			// - names matching <SECRET_NAME>-\d+
+			// This excludes all secrets with letters after the SECRET_NAME.
+			continue
+		}
+
+		// s is now a secret which either matches
+		// SECRET_NAME-*, or is exactly SECRET_NAME.
+
+		if (best.Name == "") ||
+			(len(s.Name) > len (best.Name)) ||
+			(s.Name > best.Name) {
+			// Notes on the logic above.  It assumes that
+			// the names we deal with are of the form
+			// `foo-V` where V is an integer number
+			// without leading 0's, and the `foo` is the
+			// across all considered secrets (true, foo =
+			// SECRET_NAME)
+			//
+			// A secret is take as the new best if it
+			// either the first, or its V is greater than
+			// the V of the current best.
+			//
+			// - A longer name indicates a longer V, that
+			//   is larger (*). This relies on the `no
+			//   leading 0's` part.
+			//
+			// - For V's of the same length the lexico-
+			//   graphical order is identical to the
+			//   numeric order. This also relies on the
+			//   `no leading 0's` part.
+			//
+			// (*) The special case where the name is just
+			// `foo` is handled by this as well, as a
+			// zero-length V which is less than anything
+			// else and thus only used when nothing else
+			// is present, in line with `foo` being the
+			// last resort fallback.
+
+			best = s
+		}
 	}
+
+	if best.Name != "" {
+		return &best, nil
+	}
+	return nil, errors.NewNotFound(v1.Resource("secret"), SECRET_NAME)
 }
 
 func GetOrCreateSecrets(s secretInterface) (create bool, secrets, updates *v1.Secret) {
```

Comment:
This function should become `CreateSecrets` because the versioned secrets for this revision cannot exist. There is no point in the `create` return value; it will always be true.

Reply:
Hm. I think this implies that `releaseRevision := getEnv("RELEASE_REVISION")` must never return an empty string. I.e. that I should add a fatal should happen.

The current code technically is able to handle an empty R_V, at that point it imports `secret` and writes `secret`. That is the one case where `create` can actually be `false`, still.

Reply:
> Hm. I think this implies that releaseRevision := getEnv("RELEASE_REVISION") must never return an empty string. I.e. that I should add a fatal should happen.

Agreed, this should be a fatal error. We just discussed on 🚀chat that we don't need to maintain support for unversioned secrets.

Reply:
Will do.